### PR TITLE
Congestion control API proposal

### LINKED ISSUE #206: Expose flow control information to transform
Discucssion of the late fanout use case at TPAC 2023 led to the realization that for this use case, and also for others such as "additional frame-level metadata", we need to expose congestion control signals from the transport via the packetizer to the transform, and allow them to be modified before (possibly) passing them on to the encoder (or, in the late fanout use case, to the incoming stream that it is doing fanout for).

Similarly, there may be signals that a decoder pipeline wishes to pass on to the depacketizer and to the transport, such as a keyframe request - these signals should also be exposed at the Javascript API.

A first suggestion would be to expose a series of events (one per feedback type identified) on the downstream interface of the transform, with a corresponding series of methods on the upstream interface of the transform that would take the same parameters as the content of the incoming events; this would make "default handling" easy by connecting the event directly to the handler if the transform did not desire to do any processing of them.

### PR DESCRIPTION
**Pull request by alvestrand**
Fixes #206

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-encoded-transform/pull/207.html" title="Last updated on Oct 16, 2023, 7:27 PM UTC (ee1f348)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-encoded-transform/207/09f618e...ee1f348.html" title="Last updated on Oct 16, 2023, 7:27 PM UTC (ee1f348)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -655,12 +684,25 @@ This allows algorithms to go from an {{RTCRtpScriptTransformer}} object to its {
...
+
+The <dfn attribute for="RTCRtpScriptSink">onkeyframerequest</dfn> EventHandler has type keyframerequest.
+
+## Events ##
+
+The following events fire on an RTCRtpScriptTransform:
+
+* bandwidthestimate of type {{BandwidthEstimateEvent}} - fired when the sink determines that the bandwidth estimate has changed significantly from what has been previously signalled. In the context of {{RTCRtpScriptTransform}}, if the event is not cancelled, the information will be passed to the source as if the bandwidthEstimate() function had been called.
+
+* keyframerequest of type {{KeyFrameRequestEvent}} - fired when the sink determines that a key frame has been requested. In the context of {{RTCRtpScriptTransform, if the event is not cancelled, the information will be passed to the source as if the generateKeyFrame() function had been called. <!-- NOTE IN DRAFT: generateKeyFrame() and requestKeyFrame() should merge -->
```
Comment (guidou) [inline index.bs:703]: `}}` after `RTCRtpScriptTransform`

File: index.bs [hunk 2]
```diff
@@ -586,23 +586,49 @@ interface RTCTransformEvent : Event {
...
 
-[Exposed=DedicatedWorker]
-interface RTCRtpScriptTransformer {
+interface mixin RTCRtpScriptSource {
     readonly attribute ReadableStream readable;
-    readonly attribute WritableStream writable;
-    readonly attribute any options;
     Promise&lt;unsigned long long&gt; generateKeyFrame(optional DOMString rid);
     Promise&lt;undefined&gt; sendKeyFrameRequest();
+    undefined bandwidthEstimate(BandwidthInfo info);
```
Comment (guidou) [inline index.bs:602]: This method's name looks like an accessor, but it returns undefined. Maybe add a verb before `bandwidthEstimate` (`sendBandwidthEstimate`)?

File: index.bs (hunk 1 above)
Comment (alvestrand) [reply index.bs:703]: Done!

File: index.bs (hunk 2 above)
Comment (alvestrand) [reply index.bs:602]: Done! Good catch.

**Comment by youennf** [comment]
I think there is a use case for FIR/PLI handling.
I initially thought there would be a similar need for congestion control but I wonder how much this could be dealt only by the UA monitoring the change of bandwidth induced by the transform (see also https://github.com/w3c/webrtc-encoded-transform/issues/50#issuecomment-1715996471).

I would tend to split this PR in two if we are not quickly converging on the need for the congestion control API.

**Comment by alvestrand** [comment]
Since we've had the lack of congestion control put forward as a major argument for why moving frames between peerconnections is dangerous (discussion on #201), I hope we can quickly converge on saying "yes, this is needed" and continue to "is this shape appropriate".

**Comment by youennf** [comment]
> Since we've had the lack of congestion control put forward as a major argument for why moving frames between peerconnections is dangerous

It is an issue for that specific use case, which might need a different API from the current webrtc encoded transform API.
I am unsure how much this congestion API will help developers using the current webrtc encoded transform API.

**Comment by alvestrand** [comment]
It will help for scenarios that move frames between PeerConnections (Low Latency Broadcast with Fanout, requirement N39), even though the initial deployment scenario envisioned at Google (upstream bottleneck, downstream high-capacity LAN) doesn't need it.
Agreed that a different API shape than ScriptTransform may suit this use case better, but it's possible to live within the constraints of ScriptTransform for this use case.

It will also help for the "adding lots of metadata to the encoded frame" use case, where the bandwidth that is sent on the wire is very different from the bandwidth that is produced by the encoder. That functionality can be used to satisfy requirement N23 in webrtc-nv-use-cases section 3.5

**Comment by alvestrand** [comment]
Conclusions after discussion on WEBRTC WG October interim:
- Keyframe signaling is worthwhile, but should go in a separate PR
- Signals for bandwidth changes aren't necessary; if the transform needs to look at bandwidth controls, it should just read the bandwidth info attribute every time it sends a frame

**Comment by alvestrand** [comment]
Comment on October interim resolution:
The signal for bandwidth change also served as a control surface for saying "I want to tell upstream about this myself". We should have such a knob somewhere in the API; calling the sendBandwidthInfo might be a good enough signal.

**Pull request merged**